xarray_beam/_src/pangeo_forge.py

```diff
@@ -123,25 +151,39 @@ def _open_chunks(
     """Open datasets into chunks with XArray."""
     with self._open_dataset(path) as dataset:
 
-      dataset = _expand_dimensions_by_key(dataset, index, self.pattern)
+      # We only want to expand the concat dimensions of the dataset.
+      dataset = _expand_dimensions_by_key(
+        dataset,
+        tuple((dim for dim in index if dim.name in self._concat_dims)),
+        self.pattern
+      )
+
+      if not self._max_sizes:
+        self._max_sizes = dataset.sizes
```

Comment:
Thanks, that's helpful to know that stateful transforms are discouraged; I'll keep that in mind for the future. 

This caching was the simplest way I could think of to calculate the correct offsets for the keys. See [this code-block for context](https://github.com/google/xarray-beam/pull/39/files#diff-970ec62efeaa4301061ce9fe3d12ff4ee2119ae0832eb0b52d81f3d2318b12a5R169-R172).

When I calculated the offsets using the current dataset's sizes, it would always fail to compute the last offsets correctly (please see the code-block linked above). The simplest way I could think of to calculate the right starting offset was to cache the first dataset's size, and let the 0-indexed `dim.index` handle the rest.

From what I can tell, this data is safe to cache. Those are, however, famous last words in parallel programming.